downstairs/src/region.rs

```diff
@@ -870,30 +863,50 @@ impl Region {
-
-        // Spawn parallel tasks for the flush
-        let mut join_handles: Vec<JoinHandle<Result<(), CrucibleError>>> =
-            Vec::with_capacity(extent_count);
-
-        for eid in &dirty_extents {
-            let extent = self.get_opened_extent(*eid).await;
-            let log = self.log.clone();
-            let jh = tokio::spawn(async move {
-                extent.flush(flush_number, gen_number, job_id, &log).await
-            });
-            join_handles.push(jh);
-        }
-
-        // Wait for all flushes to finish - wait until after
-        // cdt::os__flush__done to check the results and bail out.
-        let mut results = Vec::with_capacity(extent_count);
-        for join_handle in join_handles {
-            results.push(
-                join_handle
-                    .await
-                    .map_err(|e| CrucibleError::GenericError(e.to_string())),
-            );
+        // This is a bit sneaky: we want to perform each flush in a separate
+        // task for *parallelism*, but can't call `self.get_opened_extent_mut`
+        // multiple times.  In addition, we can't use the tokio thread pool to
+        // spawn a task, because that requires a 'static lifetime, which we
+        // can't get from a borrowed Extent.
+        //
+        // We'll combine two tricks to work around the issue:
+        // - Do the work in the rayon thread pool instead of using tokio tasks
+        // - Carefully walk self.extents.as_mut_slice() to mutably borrow
+        //   multiple at the same time.
+        let mut results = vec![Ok(()); dirty_extents.len()];
+        let log = self.log.clone();
+        let mut f = || {
+            let mut slice_start = 0;
+            let mut slice = self.extents.as_mut_slice();
+            let h = tokio::runtime::Handle::current();
+            rayon::scope(|s| {
+                for (eid, r) in dirty_extents.iter().zip(results.iter_mut()) {
+                    let next = eid - slice_start;
+                    slice = &mut slice[next..];
+                    let (extent, rest) = slice.split_first_mut().unwrap();
+                    let ExtentState::Opened(extent) = extent else {
+                        panic!("can't flush closed extent");
+                    };
+                    slice = rest;
+                    slice_start += next + 1;
+                    let log = log.clone();
+                    s.spawn(|_| {
+                        h.block_on(async move {
+                            *r = extent
+                                .flush(flush_number, gen_number, job_id, &log)
+                                .await
+                        });
+                    });
+                }
+            })
+        };
+        if matches!(
+            tokio::runtime::Handle::try_current().map(|r| r.runtime_flavor()),
+            Ok(tokio::runtime::RuntimeFlavor::MultiThread)
+        ) {
+            tokio::task::block_in_place(f)
```

Comment:
FWIW, in general, I find any code that uses `block_in_place` a bit sketchy, as there's a very unfortunate footgun whenever an async function that performs a `block_in_place` is used as a branch of a `tokio::select!` or a `select`-like structure (e.g. `tokio::time::timeout` etc). Since `block_in_place` blocks the calling task, if any `async fn` that uses it is part of a `select`-like structure with another future in the same task (i.e., not a spawned future), the other future being notified will _not_ wake the `select`-like structure (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=41b86b329450091574b947b08b7c89b6 for an example). This means that a function that calls `block_in_place` may cause Weird And Surprising Behavior if it's ever selected over. 

This doesn't mean it can't _ever_ be used safely, and I don't have enough context on this code to know whether that's a potential issue in this case. But, I just wanted to make sure you were aware of the potential footgun.

Reply:
Thanks for the warning!

I think that we should be safe here – we're never racing `flush` against anything else in a `select!` statement, (and shouldn't, because it's not cancel-safe!).

(Also, this is kinda a stopgap measure to avoid nuking performance with this PR; it will go away once we switch to per-extent IO tasks, which is coming up Soon™)